src/core/main.c

```diff
@@ -1166,10 +1168,24 @@ static int bump_rlimit_nofile(struct rlimit *saved_rlimit) {
                         return log_oom();
 
                 arg_default_rlimit[RLIMIT_NOFILE] = rl;
+
+                /* Bump the hard limit for system services to a substantially higher value. The default hard limit
+                 * current kernels set is pretty low (4K), mostly for historical reasons. According to kernel
+                 * developers, the fd handling in recent kernels has been optimized substantially enough, so that we
+                 * can bump the limit now, without paying too high a price in memory or performance. Note however that
+                 * we only bump the hard limit, not the soft limit. That's because select() works the way it works, and
+                 * chokes on fds >= 1024. If we'd bump the soft limit globally, it might accidentally happen to
+                 * unexpecting programs that they get fds higher than what they can process using select(). By only
+                 * bumping the hard limit but leaving the low limit as it is we avoid this pitfall: programs that are
+                 * written by folks aware of the select() problem in mind (and thus use poll()/epoll instead of
+                 * select(), the way everybody should) can explicitly opt into high fds by bumping their soft limit
+                 * beyond 1024, to the hard limit we pass. */
+                if (arg_system)
+                        arg_default_rlimit[RLIMIT_NOFILE]->rlim_max = MIN((rlim_t) nr, MAX(arg_default_rlimit[RLIMIT_NOFILE]->rlim_max, (rlim_t) HIGH_RLIMIT_NOFILE));
```

Comment:
The fact that the "global" arg_default_rlimit is used here is a bit confusing. Why not use `rl` here, and then assign at the end when the structure is ready? That way the reader immediately knows that global state is not used in the calculation:
```C
if (arg_system)
    rl->rlim_max = MIN((rlim_t) nr, MAX(rl->rlim_max, (rlim_t) HIGH_RLIMIT_NOFILE));
arg_default_rlimit[RLIMIT_NOFILE] = rl;
```